packages/eslint-plugin/src/rules/no-restricted-imports/index.js

```diff
@@ -0,0 +1,165 @@
+ * }} PathOptions
+ *
+ * @typedef {{
+ *   paths: PathOptions[];
+ * }} Options
+ */
+
+module.exports = createRule({
+  name: 'no-restricted-imports',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Restricts imports of certain packages',
+      category: 'Best Practices',
+      recommended: false,
+    },
+    messages: {
+      restrictedImport: 'Import from {{ packageName }} detected which is not allowed.',
+    },
+    fixable: 'code',
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          paths: {
+            type: 'array',
+            minItems: 1,
+            items: {
+              forbidden: {
+                type: 'array',
+                minItems: 1,
+                items: {
+                  type: 'string',
+                },
+              },
+              preferred: {
+                type: 'string',
+              },
+            },
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [],
+  create: context => {
+    /** @type {Options[]} */
+    const options = context.options;
+
+    if (!options.length) {
+      throw new Error('no-restricted-imports: Must specify a paths object.');
+    }
+
+    const paths = options[0].paths;
+    /** @type {Map<string, string | undefined>} */
+    const forbiddenPkgs = new Map();
+    /** @type {Set<string>} */
+    const preferredPkgNames = new Set();
+
+    paths.forEach(path => {
+      const { forbidden, preferred } = path;
+      forbidden.forEach(pkg => {
+        forbiddenPkgs.set(pkg, preferred);
+      });
+    });
+
+    forbiddenPkgs.forEach(preferred => {
+      if (preferred) {
+        preferredPkgNames.add(preferred);
+      }
+    });
+
+    /** @type {FixMap}*/
+    const typeImportKindFix = {};
+    /** @type {FixMap}*/
+    const valueImportKindFix = {};
+
+    return {
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      ImportDeclaration: imprt => {
+        if (!imprt.source || (imprt.source && imprt.source.type !== AST_NODE_TYPES.Literal)) {
+          return;
+        }
+
+        if (typeof imprt.source.value !== 'string') {
+          return;
+        }
+
+        const packageName = imprt.source.value;
+        const specifiers = imprt.specifiers;
+
+        if (!forbiddenPkgs.has(packageName) && !preferredPkgNames.has(packageName)) {
+          return;
+        }
+
+        /**
+         * @param {FixMap} fixMap object mapping to track if a restricted package's preferred import is
+         * already present in the file.
+         */
+        function runLintRule(fixMap) {
+          const preferredImportForCurrentPkg = forbiddenPkgs.get(packageName);
+          if (preferredPkgNames.has(packageName) && !fixMap[packageName]) {
+            fixMap[packageName] = specifiers[specifiers.length - 1];
+          }
+
+          if (forbiddenPkgs.has(packageName) && !preferredPkgNames.has(packageName)) {
+            if (
+              !preferredImportForCurrentPkg ||
```

Comment:
I don't quite understand this condition, if there is no preferred import for the current package you still add `undefined` to the fixmap ?